### PR TITLE
Maintain sorted timeslots for performance

### LINKED ISSUE #2817: Improve performance of pulse.schedule
### What is the expected enhancement?
Construction of `Schedule` using `append` should be much faster.
Currently, it takes 34s (273s) for constructing a schedule with 2000 (4000) instructions. Those are  measured by `create_random_appended_schedule(100, 20)` (200, 20) using the following code: 

```
def create_random_appended_schedule(n_instrs_by_channel, n_channels, pulse_size=10):
    channels = [pulse.DriveChannel(i) for i in range(n_channels)]
    instrs = []
    for i in range(n_instrs_by_channel):
        command = pulse.SamplePulse(np.random.random(pulse_size))
        for chan in channels:
            instrs.append(command(chan))

    start_time = time.time()
    schedule = pulse.Schedule()
    for instr in instrs:
        schedule += instr
    end_time = time.time()
    print(end_time-start_time)
```

For drastic speed up, we may need a lot of refactoring, but I think we can improve the performance somewhat by a small modification.

### PR DESCRIPTION
**Pull request by taalexander**
### Summary
Closes #2817. This PR is a follow up to #2484. Performance improves from 29.8s(236s) to 0.217s(0.519s) for the benchmark in #2817.

### Details and comments
This PR improves performance by internally storing Timeslots within TimeslotCollection in a sorted list. This allows O(1) Schedule appends, and other optimizations to be made when creating new TimeslotCollections.

Additional performance plots (different colors are a different number of channels, with same number of instructions appended for each channel).

master:
image
this PR:
image

### CODE REVIEW

File: qiskit/pulse/timeslots.py [hunk 1]
```diff
@@ -193,68 +270,163 @@ def duration(self) -> int:
...
+        insert_idx = len(ch_timeslots)
+
+        # bubble sort for insertion location.
+        # Worst case O(n_channels), O(1) for append
+        # could be improved by implementing interval tree
+        for ch_timeslot in reversed(ch_timeslots):
+            ch_interval = ch_timeslot.interval
+
+            if interval > ch_interval:
+                break
```
Comment (lcapelluto) [inline qiskit/pulse/timeslots.py]: For locality, can we move `ch_timeslots.insert(insert_idx, timeslot)` before `break`?

File: qiskit/pulse/timeslots.py [hunk 2]
```diff
@@ -193,68 +270,163 @@ def duration(self) -> int:
...
+            ch_timeslots = self._table[channel]
+            other_ch_timeslots = timeslot_collection._table[channel]
+
+            for other_ch_timeslot in other_ch_timeslots:
+                other_ch_interval = other_ch_timeslot.interval
+
+                append = True
+                for ch_timeslot in ch_timeslots:
+                    ch_interval = ch_timeslot.interval
+                    if other_ch_interval > ch_interval:
```
Comment (lcapelluto) [inline qiskit/pulse/timeslots.py]: I think you should be comparing the first interval in `other_ch_timeslots` to the _last_ interval in `ch_timeslots` if you're checking for append! Both lists of timeslots should be sorted right?

Comment (lcapelluto) [reply qiskit/pulse/timeslots.py]: 😭 This function is hard to understand.

Can you separate the checking for append out? You can also make it worst case linear time if we know the timeslots are indeed sorted.
```
for channel in common_channels:
    if ch_timeslots[-1].stop < other_ch_timeslots[0].start:
        continue  # We are appending along this channel
    
    i = 0  # iterate through this
    j = 0  # iterate through other
    while i < len(ch_timeslots) and j < len(other_ch_timeslots):
        if ch_timeslots[i].interval.has_overlap(other_ch_timeslots[j].interval):
            return False
        if ch_timeslots[i].stop <= other_ch_timeslots[j].start:
            i += 1
        else:
            j += 1
return True
```

File: qiskit/pulse/timeslots.py [hunk 3]
```diff
@@ -193,68 +270,163 @@ def duration(self) -> int:
...
+
+        insert_idx = len(ch_timeslots)
+
+        # bubble sort for insertion location.
+        # Worst case O(n_channels), O(1) for append
+        # could be improved by implementing interval tree
+        for ch_timeslot in reversed(ch_timeslots):
+            ch_interval = ch_timeslot.interval
+
+            if interval > ch_interval:
```
[the remark this answers is not in the document]
Comment (itoko) [reply qiskit/pulse/timeslots.py]: I did the same guess; `>` would compare the start times, but it is wrong. I like `if interval.start > ch_interval.stop:` and not to define `__le__` and `__ge__` for `Interval` to avoid ambiguity.

File: qiskit/pulse/schedule.py [hunk 4]
```diff
@@ -56,10 +55,10 @@ def __init__(self, *schedules: List[Union[ScheduleComponent, Tuple[int, Schedule
                 sched_timeslots = sched.timeslots
                 if insert_time:
                     sched_timeslots = sched_timeslots.shift(insert_time)
-                timeslots.append(sched_timeslots.timeslots)
+                timeslots.append(sched_timeslots)
                 _children.append(sched_pair)
 
-            self._timeslots = TimeslotCollection(*itertools.chain(*timeslots))
+            self._timeslots = TimeslotCollection(*timeslots)
```
[the remark this answers is not in the document]
Comment (itoko) [reply qiskit/pulse/schedule.py]: For locality, why don't we narrow try-except lines to the line constructing TimeslotCollection?
```
try:
    self.timeslots = TimeslotCollection(*timeslots)
except PulseError as ts_err:
...
```

File: qiskit/pulse/timeslots.py [hunk 5]
```diff
@@ -193,68 +270,163 @@ def duration(self) -> int:
...
+        Raises:
+            PulseError: If timeslots overlap
+        """
+        interval = timeslot.interval
+        ch_timeslots = self._table[timeslot.channel]
+
+        insert_idx = len(ch_timeslots)
+
+        # bubble sort for insertion location.
+        # Worst case O(n_channels), O(1) for append
```
[the remark this answers is not in the document]
Comment (itoko) [reply qiskit/pulse/timeslots.py]: This is insignificant, but this sort is more like `insert sort` than`bubble sort`. And this is also no wrong, but I think simple binary search (key=interval.start), instead of interval tree, would be enough in this case since intervals are sorted and have non-overlap.

File: qiskit/pulse/timeslots.py [hunk 6]
```diff
@@ -85,24 +85,75 @@ def shift(self, time: int) -> 'Interval':
...
+        """Whether intervals stops at value less than or equal to the
+        other interval's starting time.
+
+        Args:
+            other (Interval): other Interval
+
+        Returns:
+            bool: are self and other equal.
+        """
+        if self.stop <= other.start:
```
[the remark this answers is not in the document]
Comment (taalexander) [reply qiskit/pulse/timeslots.py]: lol

File: qiskit/pulse/timeslots.py [hunk 7]
```diff
@@ -130,6 +196,19 @@ def shift(self, time: int) -> 'Timeslot':
...
+    def has_overlap(self, other: 'Timeslot') -> bool:
+        """Check if self has overlap with `interval`.
+
+        Args:
+            other: Other Timeslot to check for overlap with
+
+        Returns:
+            bool: True if intervals overlap and are on the same channel
+        """
+        if self.interval.has_overlap(other) and self.channel == other.channel:
```
[the remark this answers is not in the document]
Comment (taalexander) [reply qiskit/pulse/timeslots.py]: My brain was malfunctioning.

File: qiskit/pulse/timeslots.py [hunk 8]
```diff
@@ -193,68 +270,163 @@ def duration(self) -> int:
...
+            ch_timeslots = self._table[channel]
+            other_ch_timeslots = other._table[channel]
+            # if channel is in self there might be an overlap
+            timeslot_idx = 0
+            if channel in common_channels:
+                for other_ch_timeslot in other_ch_timeslots:
+                    insert_idx = self._merge_timeslot(other_ch_timeslot)
+
+                    timeslot_idx += 1
+                    # timeslot was inserted at end of list. Remaining timeslots can be appended.
```
[the remark this answers is not in the document]
Comment (taalexander) [reply qiskit/pulse/timeslots.py]: The speedup is actually significant as if we are appending to large Timeslots, the behaviour is offloaded to the pure c implementation of list extension and greatly reduce the number of comparisons required by O(n). I do agree that we can remove the use of sets.

File: qiskit/pulse/timeslots.py (hunk 1 above)
Comment (taalexander) [reply qiskit/pulse/timeslots.py]: I don't think so because we need to handle the case where there no channels present.

File: qiskit/pulse/timeslots.py (hunk 2 above)
Comment (taalexander) [reply qiskit/pulse/timeslots.py]: > I think you should be comparing the first interval in other_ch_timeslots to the last interval in ch_timeslots if you're checking for append! Both lists of timeslots should be sorted right?

yes, good catch. I missed this one.

**Comment by taalexander** [comment]
Should be good to go now.

File: qiskit/pulse/timeslots.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/pulse/timeslots.py]: Ah I see. There's still a way to write this that isn't so hard to follow... the logic isn't really that complicated

[the remark this answers is not in the document]
Comment (taalexander) [reply qiskit/pulse/timeslots.py]: I'll have another stab at it and see if I can refactor while keeping the performance enhancement.

**Comment by taalexander** [comment]
Closing in favour of #2995.
